### PR TITLE
WT-12715 Fix string match for supplied zero length strings and optimize for literals.

### PR DESCRIPTION
**Pull request by ddanderson**

### CODE REVIEW

File: src/session/session_dhandle.c [hunk 1]
```diff
@@ -506,8 +506,8 @@ __wt_session_get_btree_ckpt(WT_SESSION_IMPL *session, const char *uri, const cha
      * Test for the internal checkpoint name (WiredTigerCheckpoint). Note: must_resolve is true in a
      * subset of the cases where is_unnamed_ckpt is true.
      */
-    must_resolve = WT_STRING_MATCH(WT_CHECKPOINT, cval.str, cval.len);
     is_unnamed_ckpt = cval.len >= strlen(WT_CHECKPOINT) && WT_PREFIX_MATCH(cval.str, WT_CHECKPOINT);
+    must_resolve = is_unnamed_ckpt && WT_CONFIG_LIT_MATCH(WT_CHECKPOINT, cval);
```
Comment (ddanderson) [inline src/session/session_dhandle.c:510]: I couldn't help making a tiny optimization when changing this code.  I think it's equivalent, but I'm happy to make it more like the original if it makes anyone nervous.

**Comment by ddanderson** [comment]
@sueloverso, can you please review?  It's probably best to start at src/include/misc.h, then look at the unit tests, and finally at the rest of the mechanical changes.

**Comment by sueloverso** [comment]
@ddanderson looking now. I think the PR title failure is because you added a colon at the end of the ticket number

**Comment by ddanderson** [comment]
@sueloverso, thank you!

File: src/include/misc.h [hunk 2]
```diff
@@ -333,19 +333,33 @@
 #define WT_STREQ(s, cs) (sizeof(cs) == 2 ? (s)[0] == (cs)[0] && (s)[1] == '\0' : strcmp(s, cs) == 0)
 
 /*
- * Check if a non-empty string matches a byte sequence of len bytes. The macro expects both strings
- * to be of non-zero length in order to be a match. The bytes argument does not need to be
- * null-terminated. Similar to other string related functions, this is not expected to work on NULL
- * values.
+ * Check if a literal string matches the length and content of the supplied bytes/len pair. The
+ * bytes argument does not need to be null-terminated, and it may be null if the supplied length is
+ * zero. Note this is different from the standard strncmp function, which with a supplied zero
```
Comment (sueloverso) [inline src/include/misc.h:338]: This is a good comment. I find the last sentence a bit hard to parse. Maybe something like this:
 Note this macro works differently than the standard strncmp function. When strncmp is supplied a
 zero length it returns true. When this macro gets a zero length it returns false.

File: src/include/misc.h [hunk 3]
```diff
@@ -333,19 +333,33 @@
...
+ * length, will always be true.
+ */
+#define WT_STRING_LIT_MATCH(str, bytes, len) \
```
Comment (sueloverso) [inline src/include/misc.h:341]: Nothing to change but just want to say I have a mental image of a flame when reading "LIT_MATCH".

File: src/include/misc.h [hunk 4]
```diff
@@ -333,19 +333,33 @@
...
-    return (len == 0 || (strncmp(str, bytes, len) == 0 && str[len] == '\0'));
+    return (strncmp(str, bytes, len) == 0 && str[len] == '\0');
 }
 
+/*
+ * CONFIG versions of the WT_STRING_LIT_MATCH and WT_STRING_MATCH macros. These are convenient when
+ * matching WT_CONFIG_ITEMs.
+ */
+#define WT_CONFIG_LIT_MATCH(s, cval) WT_STRING_LIT_MATCH(s, (cval).str, (cval).len)
+#define WT_CONFIG_MATCH(s, cval) WT_STRING_MATCH(s, (cval).str, (cval).len)
```
Comment (sueloverso) [inline src/include/misc.h:361]: Both for this and WT_STRING_MATCH, is it possible to write a dist script to check for literals using the wrong macro and flag them? I think that would help people remember these new ones exist. It also would help identify in case you missed any cases in this PR even.

File: src/include/misc.h [hunk 5]
```diff
@@ -333,19 +333,33 @@
...
 {
-    return (len == 0 || (strncmp(str, bytes, len) == 0 && str[len] == '\0'));
+    return (strncmp(str, bytes, len) == 0 && str[len] == '\0');
 }
 
+/*
+ * CONFIG versions of the WT_STRING_LIT_MATCH and WT_STRING_MATCH macros. These are convenient when
+ * matching WT_CONFIG_ITEMs.
+ */
+#define WT_CONFIG_LIT_MATCH(s, cval) WT_STRING_LIT_MATCH(s, (cval).str, (cval).len)
```
Comment (sueloverso) [inline src/include/misc.h:360]: I like this macro it makes things easier. We should have done this one a long time ago.

File: src/session/session_dhandle.c (hunk 1 above)
Comment (sueloverso) [reply src/session/session_dhandle.c:510]: Given that MongoDB doesn't use named checkpoints, is there any optimization if `is_unnamed_ckpt` always true? I am otherwise okay with the change as it is equivalent AFAICT.

File: src/include/misc.h [hunk 6]
```diff
@@ -333,19 +333,33 @@
...
+ * length, will always be true.
+ */
+#define WT_STRING_LIT_MATCH(str, bytes, len) \
+    ((len) == strlen("" str "") && strncmp(str, bytes, len) == 0)
+
+/*
+ * Identical to WT_STRING_LIT_MATCH, except that this works with non-literal strings. It is slightly
+ * slower, so WT_STRING_LIT_MATCH is always preferred for literal strings.
  */
 #define WT_STRING_MATCH(str, bytes, len) __wt_string_match(str, bytes, len)
```
Comment (sueloverso) [inline src/include/misc.h:348]: Is there any caller of `WT_STRING_MATCH` left? I didn't see any. I saw a bunch of `WT_CONFIG_MATCH` and everything else was literal. But it is definitely possible my eyes missed it.

File: src/include/misc.h [hunk 7]
```diff
@@ -333,19 +333,33 @@
...
 
 static WT_INLINE bool
 __wt_string_match(const char *str, const char *bytes, size_t len)
 {
-    return (len == 0 || (strncmp(str, bytes, len) == 0 && str[len] == '\0'));
+    return (strncmp(str, bytes, len) == 0 && str[len] == '\0');
```
Comment (sueloverso) [inline src/include/misc.h:353]: And if there is no called of `WT_STRING_MATCH` left, then this can go away also.

**Review by sueloverso — APPROVED** [review]
LGTM. I have several comments. The only very substantive one is it would be good to have a dist script checking for the right macro usage.

File: test/unittest/tests/test_string_match.cpp [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    empty.len = 0;
+
+    /* The "left side" of the match */
+    empty_str = "";
+    g_str = "g";
+    greed_str = "greed";
+    green_str = "green";
+    greener_str = "greener";
+    null_str = NULL;
+    something_str = "greener";
```
Comment (sueloverso) [inline test/unittest/tests/test_string_match.cpp:32]: A comment here for why `something` needs to be "greener" instead of "something" since it isn't following the pattern.

File: test/unittest/tests/test_string_match.cpp [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    if (getenv("__never__") != NULL)
+        null_str = "null";
+
+    SECTION("Check for literal string matching")
+    {
+        CHECK(null_str == NULL);
+
+        /* We run the test twice, once with the config string null terminated, one without. */
+        for (i = 0; i < 2; i++) {
+            CHECK(WT_STRING_LIT_MATCH("g", green.str, green.len) == 0);
```
Comment (sueloverso) [inline test/unittest/tests/test_string_match.cpp:47]: I think a comment here because, while WT code is often `if (WT_STRING_MATCH(...)) <do something>` is clear, these lines are confusing because `strcmp == 0` means a match and here it means doesn't match. So all the explicit checks make me really think about what is going on. Maybe even define a `MATCHES` or `DOESNT_MATCH` but that could be overkill.

File: test/unittest/tests/test_string_match.cpp [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+            green.str = "greenery";
+        }
+    }
+
+    SECTION("Check for generic string matching")
+    {
+        /* We run the test twice, once with the config string null terminated, one without. */
+        green.str = "green";
+        for (i = 0; i < 2; i++) {
+            CHECK(WT_STRING_MATCH("g", green.str, green.len) == 0);
```
Comment (sueloverso) [inline test/unittest/tests/test_string_match.cpp:80]: Per my question above, this appears to be the only caller left.

**Review by sueloverso — COMMENTED** [review]
I forgot to review the test. Here is that.

File: src/include/misc.h (hunk 6 above)
Comment (ddanderson) [reply src/include/misc.h:348]: I think there are these:
```
meta/meta_ckpt.c:          WT_STRING_MATCH(blk_mod->id_str, blk->id_str, strlen(blk->id_str))) {
schema/schema_open.c:        match = idxname == NULL || WT_STRING_MATCH(name, idxname, len);
tiered/tiered_handle.c:        if (!WT_STRING_MATCH(key, object, strlen(object)))
txn/txn_ckpt.c:            if (WT_STRING_MATCH(ckpt->name, name, len)) {
txn/txn_ckpt.c:        if (!matched && !WT_STRING_MATCH(ckpt->name, name, len))
txn/txn_ckpt.c:        if (WT_STRING_MATCH(ckpt->name, name, len))
```

File: test/unittest/tests/test_string_match.cpp (hunk 9 above)
Comment (ddanderson) [reply test/unittest/tests/test_string_match.cpp:47]: Would these be clearer as `== false` and `== true`?

File: test/unittest/tests/test_string_match.cpp (hunk 8 above)
Comment (ddanderson) [reply test/unittest/tests/test_string_match.cpp:32]: Thanks, I changed the string to `"something"`.

File: src/include/misc.h (hunk 2 above)
Comment (ddanderson) [reply src/include/misc.h:338]: Thanks, I'll use your wording!

**Comment by ddanderson** [comment]
@marcbutler can you please take a look at this change?  It's probably best to start at src/include/misc.h, then look at the unit tests, and finally at the rest of the mechanical changes.

I think I'll take @sueloverso's suggestion to add a dist script to enforce the use of the _LIT_ macro with literal strings, that will be tomorrow.

File: test/unittest/tests/test_string_match.cpp (hunk 9 above)
Comment (sueloverso) [reply test/unittest/tests/test_string_match.cpp]: Yes.

File: src/include/misc.h (hunk 6 above)
Comment (sueloverso) [reply src/include/misc.h:349]: Ah, that makes a lot of sense. Then ignore my suggestion to remove it!!

File: src/include/misc.h (hunk 4 above)
Comment (ddanderson) [reply src/include/misc.h:362]: Great suggestion for a dist script, I just pushed a change to s_function for this.  And I did miss one (and fixed it).

File: test/unittest/tests/test_string_match.cpp (hunk 9 above)
Comment (ddanderson) [reply test/unittest/tests/test_string_match.cpp]: Changed to false/true.

File: src/session/session_dhandle.c (hunk 1 above)
Comment (tod-johnson-mongodb) [reply src/session/session_dhandle.c:510]: The problem with this is someone reading it won't know the equivalence. If you do this you should add a comment saying: /* Equivalent to must_resolve = WT_STRING_MATCH(WT_CHECKPOINT, cval.str, cval.len); but faster. */

Also, I think line 510 is saying not only is WT_CHECKPOINT a prefix of cval.str, it is the same. In that case line 510 can be replaced with.

must_resolve = is_unnamed_ckptr && cval.str[strlen(WT_CHECKPOINT)] == '\0';
If the compiler replaces strlen() with the length of WT_CHECKPOINT at compile time this does not call any function.

Comment (ddanderson) [reply src/session/session_dhandle.c]: I removed this optimization, it's not perf-critical code anway.

**Comment by ddanderson** [comment]
@tod-johnson-mongodb, could you please take another look at this?  I removed the non-related optimization to keep this PR simple.

**Review by tod-johnson-mongodb — APPROVED** [review]
LGTM
